// allure-js-commons/dist/src/AllureCommandStep.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllureCommandStepExecutable = void 0;
const model_1 = require("./model");
class AllureCommandStepExecutable {
    constructor(name) {
        this.name = "";
        this.attachments = [];
        this.metadata = {};
        this.name = name;
    }
    static toExecutableItem(runtime, stepMetadata) {
        const executable = Object.assign(Object.assign({}, stepMetadata), { attachments: [], steps: [] });
        if (stepMetadata.attachments.length > 0) {
            stepMetadata.attachments.forEach((attachment) => {
                const attachmentContent = Buffer.from(attachment.content, attachment.encoding);
                const attachmentFilename = runtime.writeAttachment(attachmentContent, attachment.type, attachment.encoding);
                executable.attachments.push({
                    name: attachment.name,
                    type: attachment.type,
                    source: attachmentFilename,
                });
            });
        }
        if (stepMetadata.steps.length > 0) {
            executable.steps = stepMetadata.steps.map((nestedStep) => AllureCommandStepExecutable.toExecutableItem(runtime, nestedStep));
        }
        return executable;
    }
    label(label, value) {
        if (!this.metadata.labels) {
            this.metadata.labels = [];
        }
        this.metadata.labels.push({
            name: label,
            value,
        });
    }
    link(url, name, type) {
        if (!this.metadata.links) {
            this.metadata.links = [];
        }
        this.metadata.links.push({
            name,
            url,
            type,
        });
    }
    parameter(name, value, options) {
        if (!this.metadata.parameter) {
            this.metadata.parameter = [];
        }
        this.metadata.parameter.push({
            name,
            value,
            excluded: (options === null || options === void 0 ? void 0 : options.excluded) || false,
            mode: options === null || options === void 0 ? void 0 : options.mode,
        });
    }
    epic(epic) {
        this.label(model_1.LabelName.EPIC, epic);
    }
    feature(feature) {
        this.label(model_1.LabelName.FEATURE, feature);
    }
    story(story) {
        this.label(model_1.LabelName.STORY, story);
    }
    suite(name) {
        this.label(model_1.LabelName.SUITE, name);
    }
    parentSuite(name) {
        this.label(model_1.LabelName.PARENT_SUITE, name);
    }
    subSuite(name) {
        this.label(model_1.LabelName.SUB_SUITE, name);
    }
    owner(owner) {
        this.label(model_1.LabelName.OWNER, owner);
    }
    severity(severity) {
        this.label(model_1.LabelName.SEVERITY, severity);
    }
    tag(tag) {
        this.label(model_1.LabelName.TAG, tag);
    }
    issue(name, url) {
        this.link(url, name, model_1.LinkType.ISSUE);
    }
    tms(name, url) {
        this.link(url, name, model_1.LinkType.TMS);
    }
    attach(content, type) {
        const isBuffer = Buffer.isBuffer(content);
        this.attachments.push({
            name: "attachment",
            content: isBuffer ? content.toString("base64") : content,
            encoding: isBuffer ? "base64" : "utf8",
            type,
        });
    }
    async step(name, body) {
        if (!this.metadata.steps) {
            this.metadata.steps = [];
        }
        const nestedStep = new AllureCommandStepExecutable(name);
        const { labels = [], links = [], parameter = [], steps = [] } = await nestedStep.start(body);
        this.metadata.labels = (this.metadata.labels || []).concat(labels);
        this.metadata.links = (this.metadata.links || []).concat(links);
        this.metadata.parameter = (this.metadata.parameter || []).concat(parameter);
        this.metadata.steps = (this.metadata.steps || []).concat(steps);
    }
    async start(body) {
        const startDate = new Date().getTime();
        try {
            const res = body.call(this, this);
            await res;
            return Object.assign(Object.assign({}, this.metadata), { steps: [
                    {
                        name: this.name,
                        start: startDate,
                        stop: new Date().getTime(),
                        stage: model_1.Stage.FINISHED,
                        status: model_1.Status.PASSED,
                        statusDetails: {},
                        attachments: this.attachments,
                        parameters: [],
                        description: this.metadata.description || "",
                        descriptionHtml: this.metadata.descriptionHtml || "",
                        steps: this.metadata.steps || [],
                    },
                ] });
        }
        catch (err) {
            return Object.assign(Object.assign({}, this.metadata), { steps: [
                    {
                        name: this.name,
                        start: startDate,
                        stop: new Date().getTime(),
                        stage: model_1.Stage.FINISHED,
                        status: model_1.Status.FAILED,
                        statusDetails: err instanceof Error
                            ? {
                                message: err.message,
                                trace: err.stack,
                            }
                            : {},
                        attachments: this.attachments,
                        description: this.metadata.description || "",
                        descriptionHtml: this.metadata.descriptionHtml || "",
                        parameters: [],
                        steps: this.metadata.steps || [],
                    },
                ] });
        }
    }
}
exports.AllureCommandStepExecutable = AllureCommandStepExecutable;
